### ISSUE TITLE
Layer streaming: the VRAM pre-flight over-predicts (LOGITS_BYTES_PER_ELEMENT is stack-dependent, not preference-specific)

### ISSUE THREAD

**Issue by MakazhanAlpamys** [MakazhanAlpamys/Soup#327]
v0.72.4 budgets DPO / ORPO / SimPO at `2 x batch_size` rows, because their forward concatenates chosen and rejected into one tensor. That is correct and it never under-predicts — but it charges those rows at `ForCausalLMLoss`'s **measured 14 bytes/element**, and TRL's preference losses do not pay that.

## The gap, measured

TRL reduces logits to per-token log-probs (`selective_log_softmax`) instead of holding a full-vocabulary fp32 upcast. Measured on the dev box: DPO's entire above-resident cost was **51.76 MB**, where the 14 B/elt charge for the same shape is **~458 MB**.

Consequence on a 4 GB card (Llama-3.2-1B, vocab 128 256, NF4, 3.32 GB free):

```
seq  512  SFT 1.65 GB allowed  |  DPO 2.63 GB allowed
seq  768  SFT 2.14 GB allowed  |  DPO 3.60 GB REFUSED
seq 1024  SFT 2.63 GB allowed  |  DPO 4.58 GB REFUSED
```

So DPO is refused from `max_length: 768` upward on configurations that would probably run.

## Why no constant was published in v0.72.4

Two measurement grids were attempted and **both were invalid**, so nothing was fitted:

1. With short rows, `max_length` never bound the effective sequence — DPO measured an identical 63.43 MB at seq 128, 256 **and** 512. An independent variable that does not move measures nothing.
2. With long rows, the batch shapes showed TRL was not truncating to `max_length` either (`(1, 2401)` at every setting).

Shipping the conservative bound was the right call because under-predicting is the strictly worse failure: on Windows it is not an exception but a silent WDDM spill.

## Fix path

Get real control over the effective sequence length first (that is the actual blocker), then repeat v0.72.3's methodology: vary seq and rows at a realistic vocab, subtract the measured resident baseline, solve for the per-element slope, and use a per-loss constant.

## Acceptance criteria

- [ ] a grid where the effective sequence length demonstrably varies with the setting
- [ ] a fitted per-element cost for the preference path, with worst-case error stated
- [ ] the estimator still **never** under-predicts across the grid
- [ ] the seq-768 DPO configuration above is no longer refused

**Comment by MakazhanAlpamys** [comment]
## Measured on real models — and the cause is not the preference path

72 runs, 3 repeats per cell, **Qwen2.5-0.5B-Instruct (vocab 151936)**. The earlier two attempts were invalid because the fixture was too small — DPO measured an identical 63.43 MiB at seq 128, 256 **and** 512, an independent variable that never moved. Every run here records the tensor shape the model actually saw; `observed_rows x observed_seq` tracks `2·batch x max_length` (DPO) and `batch x max_length` (SFT) in all 72.

Spread is **exactly 0.00 MiB in every cell** — `max_memory_allocated` is bit-reproducible for a fixed shape, so the repeats confirm determinism rather than estimate noise.

### The bug, one row

| batch | seq | predicted | measured | refused @3.32 GB free |
|---|---|---|---|---|
| 1 | 768 | 3541.84 MiB | **3092.46 MiB** | **REFUSED — but it fits, with 73.74 MiB to spare** |

### It is not preference-specific

The SFT control over-predicts by the same amount at the same *effective* row count: DPO b1 vs SFT b2, DPO b2 vs SFT b4 agree to **+0.09%…+0.51%** (a flat ~6.5 MiB) across all 8 comparable shapes. So `_STREAM_ROWS_PER_EXAMPLE = 2` is **correct** — a streamed DPO step costs what a streamed SFT step at twice the rows costs.

What is wrong is one coefficient on the **shared** logits term. The ratio rises and asymptotes (DPO 1.107 → 1.167, SFT 0.987 → 1.163), the signature of a slope error plus a small offset. Fitting `measured = a + b·(vocab x seq x rows)` gives **b = 12.311 bytes/element**, worst residual 0.15% over 12 DPO cells, against the shipped `LOGITS_BYTES_PER_ELEMENT = 14`.

### Deliberately not changed yet, and why

14 was measured on RTX 3050 / Windows / an older torch; 12.3 is this stack. v0.72.3's own grid — 10 real runs, worst error 0.85%, never under-predicting — still passes unchanged. Swapping the constant on the strength of one stack would trade a **documented over-prediction** for an **undocumented under-prediction**, and on Windows/WDDM that does not raise, it silently spills into host memory.

So the honest framing is: the constant is stack-dependent, and the fix needs either a per-stack measurement or a term that does not depend on it. Retitling this issue away from "preference losses" would be reasonable.

Method note: peak must be reset **after** `setup()` — a peak spanning it charges the pre-flight's own 4096³ GEMM probe to the training step and reports an under-prediction that does not exist.

Robustness: 30 cells re-run against the tree carrying the #331 fix gave **byte-identical** peaks, and the SFT 1x256 cell reproduces the −1.3% under-prediction already documented as a known −1.7% SFT miss.

Record: `benchmarks/gate-h100-validation.md` STEP 16.

**Comment by MakazhanAlpamys** [comment]
## The mechanism, and why the constant is NOT being lowered

**14 = 12 + 2, and only the 2 differs between stacks.** Measured stage by stage on an H100, 3 repeats, spread exactly 0 in every cell:

| after logits alloc | loss-forward peak | after the loss returns | backward peak |
|---|---|---|---|
| 2.0000 | 10.0000 | 6.0000 | **14.0000** |

The peak holds the bf16 logits **plus three fp32 logits-shaped buffers**. Note this **corrects v0.72.3's recorded decomposition** ("bf16 + fp32 upcast + log-softmax + grad"): right in total, wrong in detail — the fp32 upcast is freed when the function returns.

Isolating the loss arithmetic gives **12.0000–12.0960 B/elt** across vocab {32000, 128256, 151936} × tokens {512, 1024, 2048}.

**A single-variable control names the remaining 2:** holding the model output object across `backward()` instead of letting it die with the local — nothing else changed — costs **exactly +2.0000 B/elt**.

### The loss path is already stack-independent

| stack | isolated loss path | real trl training step |
|---|---|---|
| torch 2.13.0+cu130 / trl 0.26.2 | 12.0000 | 12.0955 / 12.0854 |
| torch 2.5.1+cu124 / trl 0.19.1 (the v0.72.3 versions, installed on the box to check) | 12.0000 | **12.0955 / 12.0854** |

Byte-identical. **torch, trl and transformers are exonerated.** What differs between the two published grids is a **retention** — and a synthetic pre-flight probe structurally cannot observe it, because it sees its own reference, not whether the training loop that will run still holds the logits when the loss backward peaks. That retention could not be reproduced here, so it was not modelled away.

### The counterfactual that settles it

| slope | rows under-predicted (of the 10 v0.72.3 real runs) | worst error |
|---|---|---|
| **14 (shipped)** | **0 / 10** | +0.85% (over) |
| 12.311 (the H100 fit from the grid above) | **10 / 10** | **−10.49%** |
| 12 (measured loss path) | 10 / 10 | −12.58% |

**So this issue's acceptance criterion — seq-768 DPO no longer refused — is deliberately NOT met.** Meeting it requires a slope that under-predicts 10 of 10 of the other stack's real measurements, and on Windows/WDDM an under-prediction does not raise; it silently spills into host memory.

### What shipped instead (`5e356ae`)

- the constant is split into two independently measured terms that **sum to the unchanged 14**
- an **opt-in, upward-only** calibration `max(14, measured + 2)`, ceil so a fractional reading never rounds down. This closes a real unguarded hole: a future stack that grew a fourth fp32 buffer would be under-budgeted by 12.5% today with nothing to catch it.
- default behaviour byte-identical; no new CUDA on the pre-flight path

**Staying open** because the user-visible symptom — a refusal of configs that would fit — is still there. What it needs is either a way to observe the retention at pre-flight time, or a decision to accept a documented over-prediction as the price of never spilling. That is a product call, not a measurement one.

Record: `benchmarks/gate-h100-validation.md` STEP 16.

**Comment by MakazhanAlpamys** [comment]
## Decision: `LOGITS_BYTES_PER_ELEMENT` stays at 14. Deliberately.

Three options were weighed against what STEP 16 of
`benchmarks/gate-h100-validation.md` actually measured, plus the mechanism pass
that decomposed the constant into `12 + 2`.

**Keep 14 (chosen).** It is the only value under which *neither* published grid
under-predicts. Counterfactual, over the 10 real runs in
`tests/test_v07203.py::TestPeakVramReproducesTheMeasuredGrid`:

| slope | under-predicts | worst error |
|---|---|---|
| **14, shipped** | **0 / 10** | +0.85%, over |
| 12.311 (this-box fit) | 10 / 10 | −10.49% |
| 12 (loss path alone) | 10 / 10 | −12.58% |

**Lower it to the measured value — rejected.** It closes this issue by the letter
and reopens a worse one: 10 of 10 rows of the published grid start under-predicting,
and on Windows/WDDM an under-prediction is not an exception — it silently spills to
host memory (measured: 9.27 GB allocated on a 4.29 GB card, no exception raised).
That also quietly retires the "peak is bounded by one layer" claim. Changing the
constant would require deleting or invalidating the only real-hardware validation
the estimator has.

**Split the constant per platform — rejected.** There are two data points and they
confound OS, GPU, driver and session. The record states plainly that the source of
the retention is **not identified**, and that torch, trl and transformers were each
swapped to the other grid's versions and measured byte-identical, so none of them is
the cause. A `if platform == "win32"` branch would encode a causal claim no
measurement supports — the same class of error STEP 26 of that record corrects
elsewhere (a sound observation carrying an unsound attribution). That the
correlation happens to point in the safe direction is a coincidence of n=1 per
platform, not a defence.

## So this issue's symptom stays, and is closed elsewhere

The asymmetry justifies the **default**, not the absence of an exit. Today a refusal
is a wall. Follow-ups filed:

- **#347 — `training.stream_vram_override`.** An explicit, warned opt-out of the
  refusal. Default stays 14, no grid row is touched, and the estimator keeps its
  never-under-predicts property; only the *action* on a predicted miss becomes the
  user's call. **This is the substantive closure of this issue** — the symptom is
  "a config that would fit is refused with no way through", and this removes the
  second half without trading a documented over-prediction for an undocumented
  under-prediction.
- **#348 — `calibrated_logits_bytes_per_element` is dead code.** Found while
  deciding this: the upward-only calibration described as shipped alongside the
  split constant has **no caller in `src/`**. `stream_setup.py:455` never passes
  `logits_bytes_per_element=`, and there is no flag or env var reaching it. The hole
  it was written to close — a future stack growing a fourth fp32 buffer, 12.5%
  under-budgeted with no guard — is still open. The wording in the gate record is
  being corrected to "an API, not yet wired into the pre-flight".
- **#349 — measure instead of predict.** The retention is invisible to a synthetic
  probe by construction. Running one real forward+backward at the configured shape
  after `setup()` sees it, because it *is* the real loop. That is the path that ends
  the constant debate rather than re-tuning it.

Keeping this issue open, with the acceptance criterion (seq-768 DPO no longer
refused) explicitly **not met by design**, until #347 lands.

**Comment by MakazhanAlpamys** [comment]
A third independent measurement of this constant, and the first not taken on the H100.

Method is different from both previous ones, which is what makes it worth adding: instead of instrumenting the loss, it reads the constant off the **marginal cost of one micro-batch row**, with and without Cut Cross-Entropy removing the logits materialisation. Llama-3.1-8B NF4, streamed, S=512, vocab 128 256, RTX 3050 Laptop.

| micro-batch | peak, no CCE | peak, CCE |
|---|---|---|
| 1 | 3.448 GB | 3.144 GB |
| 2 | 4.512 GB | 3.463 GB |
| 3 | 5.573 GB | 3.782 GB |

Marginal cost per row: **1.063 GB** without CCE, **0.319 GB** with (both linear to three decimals across the two intervals). The difference is the logits term and nothing else:

```
0.744 GB / (512 x 128256) = 11.32 bytes per element
```

against the **14** charged today. It lands close to the H100 decomposition's *loss-arithmetic* term of 12 and clearly below 14, which is consistent with the retention term (+2) not appearing in a marginal-per-row measurement.

This does **not** change the recommendation to keep 14. The asymmetry that decided it still holds — over-predicting refuses a config that would have worked, under-predicting on Windows/WDDM raises nothing and silently spills — and a third method landing near 12 does not make 12 safe as a floor.

What it does add is a bound on how much is being left on the table at this configuration, measured rather than modelled, plus a concrete interaction: with `training.use_cut_ce` on, the marginal logits cost is ~0 and the pre-flight would refuse configurations that in fact fit. That flag currently cannot engage at all (#383), so nothing is broken today, but the two issues have to be resolved together.

Full record and method: `benchmarks/probe-v0.73.0-what-bounds-streaming.md`.

**Comment by MakazhanAlpamys** [comment]
Cross-reference from v0.73.1: this issue's framing — that the pre-flight **over**-predicts — turns out to be regime-specific.

Measured through the real `soup train` on the same box, the formula over-predicts at short sequence (1.081x at seq 4352) and **under**-predicts past that: 0.934x the real peak at seq 5120 and 0.787x at 6144. So `LOGITS_BYTES_PER_ELEMENT = 14` being conservative is true where this issue measured it and false at long sequence, for a reason that is not the logits constant at all.

Filed separately as #395, since the cause and the direction both differ. The decision recorded here — constant stays at 14, upward-only calibration — is unaffected and still correct for the regime it was taken in.

Record: [`benchmarks/gate-v0.73.1-measured-vram-fit.md`](https://github.com/MakazhanAlpamys/Soup/blob/v0.73.1/benchmarks/gate-v0.73.1-measured-vram-fit.md)

**Comment by MakazhanAlpamys** [comment]
Closing. On 2026-08-08 I said this stays open until #347 lands; it landed on 2026-08-13 and I did not follow through.

What resolved the symptom this issue was about (a streamed config that would fit, refused with no way through):
- `training.stream_vram_override` (#347), by @AmirF194 in #386: an explicit, warned replacement for the measured free VRAM, usable in either direction.
- The upward-only calibration wired into the pre-flight (#348), by @AmirF194 in #390.
- `training.stream_vram_probe` (#349, v0.73.1), which measures one real step at the configured shape instead of predicting it (SFT only).

`LOGITS_BYTES_PER_ELEMENT` stays at 14, as decided above. What remains is tracked elsewhere: a probe that measures the DPO/ORPO/SimPO/KTO step itself is #840, the under-prediction at long sequence is #395, and the `use_cut_ce` interaction from my 08-11 comment is now #1206.
